Add fallback code for launching WebKit process extensions

pvollan opened the pull request:
#### 5ca67b41a22f9c7bee0aafa7a6669c710881c7e9
<pre>
Add fallback code for launching WebKit process extensions
<a href="https://bugs.webkit.org/show_bug.cgi?id=270238">https://bugs.webkit.org/show_bug.cgi?id=270238</a>
<a href="https://rdar.apple.com/123774804">rdar://123774804</a>

Reviewed by Timothy Hatcher and Chris Dumez.

Add fallback code for launching WebKit process extensions. This is required to run layout tests in Simulator with WebKit ToT
in certain older configurations. This fallback code is only needed on Simulator and only enabled there.

* Source/WebKit/Platform/cocoa/ExtensionCapability.h:
(WebKit::ExtensionCapability::platformCapability const):
(WebKit::ExtensionCapability::hasPlatformCapability const):
(WebKit::ExtensionCapability::platformCapabilityIsValid):
(WebKit::ExtensionCapability::setPlatformCapability):
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.h:
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
(WebKit::platformInvalidate):
(WebKit::ExtensionCapabilityGrant::ExtensionCapabilityGrant):
(WebKit::ExtensionCapabilityGrant::~ExtensionCapabilityGrant):
(WebKit::ExtensionCapabilityGrant::isEmpty const):
(WebKit::ExtensionCapabilityGrant::isValid const):
(WebKit::ExtensionCapabilityGrant::setPlatformGrant):
(WebKit::ExtensionCapabilityGrant::invalidate):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/GPUExtension-Info.plist:
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-CaptivePortal-Info.plist:
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-Info.plist:
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
(WebKit::grantCapability):
(WebKit::grantCapabilityInternal):
(WebKit::ExtensionCapabilityGranter::grant):
(WebKit::ExtensionCapabilityGranter::setMediaCapabilityActive):
(WebKit::ExtensionCapabilityGranter::invalidateGrants):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::acquireSync):
(WebKit::ProcessAssertion::~ProcessAssertion):
* Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
(WebKit::ExtensionProcess::ExtensionProcess):
(WebKit::ExtensionProcess::grantCapability const):
(WebKit::ExtensionProcess::createVisibilityPropagationInteraction const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::launchWithExtensionKitFallback):
(WebKit::launchWithExtensionKit):
* Source/WebKit/UIProcess/ProcessAssertion.h:

Canonical link: <a href="https://commits.webkit.org/275590@main">https://commits.webkit.org/275590@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/265927a4d03d8eb73f81cd0f3d5c153f034b876b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42234 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21252 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44628 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44829 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38350 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24453 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18586 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34983 "Failure limit exceed. At least found 140 new test failures: accessibility/math-multiscript-attributes.html, animations/keyframe-em-unit.html, animations/shadow-host-child-change.html, compositing/canvas/accelerated-canvas-compositing-size-limit.html, compositing/tiling/sticky-change-to-tiled.html, css1/font_properties/font.html, css1/pseudo/multiple_pseudo_elements.html, css2.1/20110323/absolute-non-replaced-height-007.htm, css2.1/20110323/absolute-non-replaced-height-009.htm, css2.1/20110323/absolute-non-replaced-max-height-002.htm ... (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42808 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18188 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36372 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15920 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15850 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37414 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46277 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/35690 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38429 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37742 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41643 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/41862 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17050 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14035 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40223 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18669 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/48871 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18731 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9924 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5693 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18314 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/Platform/cocoa/ExtensionCapability.h:
```diff
@@ -31,21 +31,34 @@
 protected:
     ExtensionCapability() = default;
-    void setPlatformCapability(BEProcessCapability *capability) { m_platformCapability = capability; }
+    void setPlatformCapability(PlatformCapability capability) { m_platformCapability = capability; }
```
Should this be `PlatformCapability&&` since we are dealing with `RetainPtr`'s now?

cdumez replied:
`PlatformCapability&&` or `const PlatformCapability&`. Also `std::move()` if you go with the r-value reference.

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -35,14 +35,21 @@
-        RELEASE_LOG_ERROR(ProcessCapabilities, "Invalidating grant %{public}@ failed", platformGrant);
+    WTF::switchOn(platformGrant, [&] (auto& grant) {
+        if (![grant isValid])
+            return;
```
This returns from the lambda function, not from the platformInvalidate() function. As a result, your switchOn() here is a no-op and we no longer early return when the grant isn't valid.

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -72,21 +79,30 @@ static void platformInvalidate(id<BEProcessCapabilityGrant> platformGrant)
 bool ExtensionCapabilityGrant::isEmpty() const
 {
-    return !m_platformGrant;
+    WTF::switchOn(m_platformGrant, [&] (auto& grant) {
```
Again, this whole switchOn() is a no-op currently.

This should be:
```
return WTF::switchOn(m_platformGrant, [&] (auto& grant) {
    return !grant;
});
```

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -72,21 +79,30 @@ static void platformInvalidate(id<BEProcessCapabilityGrant> platformGrant)
 #if USE(EXTENSIONKIT)
-    if ([m_platformGrant isValid])
-        return true;
+    WTF::switchOn(m_platformGrant, [&] (auto& grant) {
```
Same bug.

cdumez reviewed (changes requested):
Not done reviewing but already setting r- to make sure this doesn't land given this PR got approved and has significant bugs.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -71,20 +75,60 @@
+{
+    auto [name, identifier] = serviceNameAndIdentifier(processType, client, processLauncher.isRetryingLaunch());
+
+    auto configuration = adoptNS([alloc_SEServiceConfigurationInstance() initWithServiceIdentifier:identifier.get()]);
```
auto -> RetainPtr.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -71,20 +75,60 @@
     }
 }
 
+static void launchWithExtensionKitFallback(ProcessLauncher& processLauncher, ProcessLauncher::ProcessType processType, ProcessLauncher::Client* client, std::function<void(ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, ExtensionProcess&& process, ASCIILiteral name, NSError *error)> handler)
```
Why is this using std::function instead of WTF::Function? Also, why is it passed by value?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -71,20 +75,60 @@
+    }
+}
+
 static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLauncher::ProcessType processType, ProcessLauncher::Client* client, std::function<void(ThreadSafeWeakPtr<ProcessLauncher> weakProcessLauncher, ExtensionProcess&& process, ASCIILiteral name, NSError *error)> handler)
```
Why is this using std::function instead of WTF::Function? Also why is it being passed by value?

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -51,15 +58,15 @@ static void platformInvalidate(id<BEProcessCapabilityGrant> platformGrant)
 ExtensionCapabilityGrant::~ExtensionCapabilityGrant()
 {
-    setPlatformGrant(nil);
+    platformInvalidate(m_platformGrant);
```
Shouldn't this be `platformInvalidate(std::exchange(m_platformGrant, nullptr));`?

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -72,21 +79,30 @@ static void platformInvalidate(id<BEProcessCapabilityGrant> platformGrant)
+void ExtensionCapabilityGrant::invalidate()
 {
-    platformInvalidate(std::exchange(m_platformGrant, WTFMove(platformGrant)).get());
+    platformInvalidate(m_platformGrant);
```
Shouldn't this use `std::exchange(m_platformGrant, nullptr)`?

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -35,14 +35,21 @@
 namespace WebKit {
 
-static void platformInvalidate(id<BEProcessCapabilityGrant> platformGrant)
+static void platformInvalidate(PlatformGrant platformGrant)
```
This is a variant containing RetainPtr, it should not be passed by value. Maybe an r-value would be appropriate here. If not, a const reference.

pvollan replied:
Will fix!

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapability.h:
```diff
@@ -31,21 +31,34 @@
+
+    static bool platformCapabilityIsValid(const PlatformCapability& platformCapability)
+    {
+        WTF::switchOn(platformCapability, [&] (auto& capability) {
```
```
return WTF::switchOn(platformCapability, [&] (auto& capability) {
    return !!capability;
});
```

Your current code always returns false. How was this code tested?

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapability.h:
```diff
@@ -31,21 +31,34 @@
     virtual ~ExtensionCapability() = default;
     virtual String environmentIdentifier() const = 0;
-    BEProcessCapability *platformCapability() const { return m_platformCapability.get(); }
+    PlatformCapability platformCapability() const { return m_platformCapability; }
```
Should return a `const PlatformCapability&`

pvollan replied:
Ah, good catch, will fix!

cdumez commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -52,19 +52,18 @@
 
 #if USE(EXTENSIONKIT)
-static RetainPtr<BEProcessCapabilityGrant> grantCapability(BEProcessCapability *capability, const std::optional<ExtensionProcess>& process)
+static PlatformGrant grantCapability(PlatformCapability capability, const std::optional<ExtensionProcess>& process)
```
`PlatformCapability` shouldn't be passed by value.

cdumez commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -52,19 +52,18 @@
-    if (!capability || !process)
-        return nil;
+    if (!ExtensionCapability::platformCapabilityIsValid(capability) || !process)
+        return PlatformGrant();
```
nit: could just `return { };`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -69,8 +75,41 @@
     NSError *error = nil;
     RetainPtr<BEProcessCapabilityGrant> grant;
-    WTF::switchOn(m_process, [&] (auto& process) {
+    WTF::switchOn(m_process, [&] (RetainPtr<BEWebContentProcess> process) {
```
You don't need so many lambdas as far as I know. As I told you last time, you can rely on `auto`:
```
WTF::switchOn(m_process, [&] (auto& process) {
    grant = [process grantCapability:capability error:&error];
}, [](const RetainPtr<_SEExtensionProcess>&) { });
```

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -69,8 +75,41 @@
+{
+    NSError *error = nil;
+    PlatformGrant grant;
+    WTF::switchOn(m_process, [&] (RetainPtr<BEWebContentProcess> process) {
```
Same thing here, can't this use am `auto&` lambda parameter? Seems like most lambdas are identical

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -83,6 +122,7 @@
     }, [&] (RetainPtr<BENetworkingProcess> process) {
     }, [&] (RetainPtr<BERenderingProcess> process) {
         interaction = [process createVisibilityPropagationInteraction];
+    }, [&] (RetainPtr<_SEExtensionProcess> process) {
```
Could use `auto&` here and drop the `RetainPtr<BENetworkingProcess> process` lambda above.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -71,20 +75,60 @@
+    case ProcessLauncher::ProcessType::Web: {
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SEExtensionProcess *_Nullable process, NSError *_Nullable error) {
+            ExtensionProcess extensionProcess(process);
+            handler(weakProcessLauncher, WTFMove(extensionProcess), name, error);
```
Why are we passing the `weakProcessLauncher` by value here and below?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -71,20 +75,60 @@
+    case ProcessLauncher::ProcessType::Network: {
+        auto block = [handler = handler, weakProcessLauncher = ThreadSafeWeakPtr { processLauncher }, name = name](_SEExtensionProcess *_Nullable process, NSError *_Nullable error) {
+            ExtensionProcess extensionProcess(process);
+            handler(weakProcessLauncher, WTFMove(extensionProcess), name, error);
```
`WTFMove(extensionProcess)` is almost the same length as `ExtensionProcess(process)`. But also, I suspect passing `process` would just work since the `ExtensionProcess` constructor is not marked as explicit. The same comment applies below.

brentfulgham reviewed (changes requested):
Just clearly documenting that I had asked Per Arne to revise things to be conditionally compiled so that the SPI is only used for the simulator.

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapability.h:
```diff
@@ -31,21 +31,43 @@
 
 namespace WebKit {
 
+#if PLATFORM(IOS_SIMULATOR)
```
Should this be gated on an OS version too? Surely we want to use the proper API on recent SDKs?

pvollan replied:
Yes, very good point, will fix!

Thanks for reviewing!

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapability.h:
```diff
@@ -31,21 +31,43 @@
+    static bool platformCapabilityIsValid(const PlatformCapability& platformCapability)
+    {
+#if PLATFORM(IOS_SIMULATOR)
+        return WTF::switchOn(platformCapability, [&] (auto& capability) {
```
Capturing `&` seems unnecessary

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapability.h:
```diff
@@ -31,21 +31,43 @@
 protected:
     ExtensionCapability() = default;
-    void setPlatformCapability(BEProcessCapability *capability) { m_platformCapability = capability; }
+    void setPlatformCapability(const PlatformCapability& capability) { m_platformCapability = capability; }
```
Could this be an r-value reference with `WTFMove()` by any chance?

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -35,31 +35,44 @@
         return;
 
-#if USE(EXTENSIONKIT)
+    WTF::switchOn(platformGrant, [&] (RetainPtr<BEProcessCapabilityGrant> grant) {
```
`const RetainPtr<BEProcessCapabilityGrant>&` to avoid ref-counting churn.

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -35,31 +35,44 @@
+    WTF::switchOn(platformGrant, [&] (RetainPtr<BEProcessCapabilityGrant> grant) {
+        if (![grant invalidate])
+            RELEASE_LOG_ERROR(ProcessCapabilities, "Invalidating grant %{public}@ failed", grant.get());
+    }, [&] (RetainPtr<_SEGrant> grant) {
```
ditto.

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -72,21 +85,34 @@ static void platformInvalidate(id<BEProcessCapabilityGrant> platformGrant)
 bool ExtensionCapabilityGrant::isEmpty() const
 {
+#if PLATFORM(IOS_SIMULATOR)
+    return WTF::switchOn(m_platformGrant, [&] (auto& grant) {
```
Unnecessary capture of `&`

cdumez commented on Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
```diff
@@ -72,21 +85,34 @@ static void platformInvalidate(id<BEProcessCapabilityGrant> platformGrant)
-    if ([m_platformGrant isValid])
-        return true;
+#if PLATFORM(IOS_SIMULATOR)
+    return WTF::switchOn(m_platformGrant, [&] (auto& grant) {
```
ditto

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.h:
```diff
@@ -33,22 +36,33 @@ OBJC_CLASS BEWebContentProcess;
     void invalidate() const;
     OSObjectPtr<xpc_connection_t> makeLibXPCConnection() const;
     RetainPtr<BEProcessCapabilityGrant> grantCapability(BEProcessCapability *) const;
+    PlatformGrant grantCapability(PlatformCapability) const;
```
Should not pass by value.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -71,7 +79,32 @@
     RetainPtr<BEProcessCapabilityGrant> grant;
     WTF::switchOn(m_process, [&] (auto& process) {
         grant = [process grantCapability:capability error:&error];
+    }, [&] (RetainPtr<_SEExtensionProcess> process) {
```
Doesn't need to capture `&`. Also, does the compiler require us to specify `process`?

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -71,7 +79,32 @@
+    PlatformGrant grant;
+#if PLATFORM(IOS_SIMULATOR)
+    WTF::switchOn(m_process, [&] (auto& process) {
+        WTF::switchOn(capability, [&] (RetainPtr<BEProcessCapability> capability) {
```
`const RetainPtr<BEProcessCapability>&` to avoid refcounting churn.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -71,7 +79,32 @@
+    WTF::switchOn(m_process, [&] (auto& process) {
+        WTF::switchOn(capability, [&] (RetainPtr<BEProcessCapability> capability) {
+            grant = [process grantCapability:capability.get() error:&error];
+        }, [&] (RetainPtr<_SECapability>) {
```
unnecessary capture of `&`. Also `const RetainPtr<_SECapability>&`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -71,7 +79,32 @@
+            grant = [process grantCapability:capability.get() error:&error];
+        }, [&] (RetainPtr<_SECapability>) {
+        });
+    }, [&] (RetainPtr<_SEExtensionProcess> process) {
```
`const RetainPtr<_SEExtensionProcess>&`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -71,7 +79,32 @@
+        }, [&] (RetainPtr<_SECapability>) {
+        });
+    }, [&] (RetainPtr<_SEExtensionProcess> process) {
+        WTF::switchOn(capability, [&] (RetainPtr<BEProcessCapability>) {
```
Unnecessary capture of `&`. `const RetainPtr<BEProcessCapability>&`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -71,7 +79,32 @@
+        });
+    }, [&] (RetainPtr<_SEExtensionProcess> process) {
+        WTF::switchOn(capability, [&] (RetainPtr<BEProcessCapability>) {
+        }, [&] (RetainPtr<_SECapability> capability) {
```
`const RetainPtr<_SECapability>&`

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -80,9 +113,9 @@
-    }, [&] (RetainPtr<BENetworkingProcess> process) {
     }, [&] (RetainPtr<BERenderingProcess> process) {
         interaction = [process createVisibilityPropagationInteraction];
+    }, [&] (auto& process) {
```
Unnecessary capture of `&`

cdumez reviewed (changes requested):
Still needs a few changes.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
```diff
@@ -71,7 +79,32 @@
     RetainPtr<BEProcessCapabilityGrant> grant;
     WTF::switchOn(m_process, [&] (auto& process) {
         grant = [process grantCapability:capability error:&error];
+    }, [] (RetainPtr<_SEExtensionProcess>) {
```
`const RetainPtr<_SEExtensionProcess>&`

pvollan replied:
Fixed!

Thanks for reviewing!

cdumez reviewed (approved):
LGTM with nit.

pvollan committed:
Add fallback code for launching WebKit process extensions
https://bugs.webkit.org/show_bug.cgi?id=270238
rdar://123774804

Reviewed by Timothy Hatcher and Chris Dumez.

Add fallback code for launching WebKit process extensions. This is required to run layout tests in Simulator with WebKit ToT
in certain older configurations. This fallback code is only needed on Simulator and only enabled there.

* Source/WebKit/Platform/cocoa/ExtensionCapability.h:
(WebKit::ExtensionCapability::platformCapability const):
(WebKit::ExtensionCapability::hasPlatformCapability const):
(WebKit::ExtensionCapability::platformCapabilityIsValid):
(WebKit::ExtensionCapability::setPlatformCapability):
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.h:
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
(WebKit::platformInvalidate):
(WebKit::ExtensionCapabilityGrant::ExtensionCapabilityGrant):
(WebKit::ExtensionCapabilityGrant::~ExtensionCapabilityGrant):
(WebKit::ExtensionCapabilityGrant::isEmpty const):
(WebKit::ExtensionCapabilityGrant::isValid const):
(WebKit::ExtensionCapabilityGrant::setPlatformGrant):
(WebKit::ExtensionCapabilityGrant::invalidate):
* Source/WebKit/Shared/AuxiliaryProcessExtensions/GPUExtension-Info.plist:
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-CaptivePortal-Info.plist:
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-Info.plist:
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
(WebKit::grantCapability):
(WebKit::grantCapabilityInternal):
(WebKit::ExtensionCapabilityGranter::grant):
(WebKit::ExtensionCapabilityGranter::setMediaCapabilityActive):
(WebKit::ExtensionCapabilityGranter::invalidateGrants):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::acquireSync):
(WebKit::ProcessAssertion::~ProcessAssertion):
* Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ExtensionProcess.mm:
(WebKit::ExtensionProcess::ExtensionProcess):
(WebKit::ExtensionProcess::grantCapability const):
(WebKit::ExtensionProcess::createVisibilityPropagationInteraction const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::launchWithExtensionKitFallback):
(WebKit::launchWithExtensionKit):
* Source/WebKit/UIProcess/ProcessAssertion.h:

Canonical link: https://commits.webkit.org/275590@main
